frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx

```diff
@@ -263,50 +248,213 @@ function DataFrame({
+          {
+            id: element.id,
+            formId: element.formId,
+          } as WidgetInfo,
+          newWidgetState,
+          {
+            fromUi: true,
+          },
+          fragmentId
+        )
       }
+    }),
+    [element.id, element.formId, widgetMgr, fragmentId]
+  )
 
-      // Use debounce to prevent rapid updates to the widget state.
-      debounce(DEBOUNCE_TIME_MS, () => {
-        const currentEditingState = editingState.current.toJson(columns)
-        let currentWidgetState = widgetMgr.getStringValue(
-          element as WidgetInfo
-        )
+  const {
+    gridSelection,
+    isRowSelectionActivated,
+    isMultiRowSelectionActivated,
+    isColumnSelectionActivated,
+    isMultiColumnSelectionActivated,
+    isRowSelected,
+    isColumnSelected,
+    isCellSelected,
+    clearSelection,
+    processSelectionChange,
+  } = useSelectionHandler(element, isEmptyTable, disabled, syncSelectionState)
 
-        if (currentWidgetState === undefined) {
-          // Create an empty widget state
-          currentWidgetState = new EditingState(0).toJson([])
-        }
+  React.useEffect(() => {
+    // Clear cell selections if fullscreen mode changes
+    // but keep row & column selections.
+    // In the past we saw some weird side-effects, so we decided to clean
+    // it when entering fullscreen-mode. If we decide to change this, we have
+    // to play around and get to the bottom of it.
+    clearSelection(true, true)
+    // Only run this on changes to the fullscreen mode:
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [isFullScreen])
 
-        // Only update if there is actually a difference between editing and widget state
-        if (currentEditingState !== currentWidgetState) {
-          widgetMgr.setStringValue(
-            element as WidgetInfo,
-            currentEditingState,
-            {
-              fromUi: triggerRerun,
-            },
-            fragmentId
-          )
+  // This callback is used to refresh the rendering of specified cells
+  const refreshCells = React.useCallback(
+    (
+      cells: {
+        cell: GridCellPosition
+      }[]
+    ) => {
+      dataEditorRef.current?.updateCells(cells)
+    },
+    []
+  )
+
+  /**
+   * On the first rendering, try to load initial selection state
+   * from the widget state if it exists. This is required in the
+   * case that other elements are inserted before this widget.
+   *
+   * This effect needs to run after the fullscreen effect that
+   * clears cell selections, since both modify the same state object.
+   */
+  React.useEffect(
+    () => {
+      if (!isRowSelectionActivated && !isColumnSelectionActivated) {
+        // Only run this if selections are activated.
+        return
+      }
+
+      const initialWidgetValue = widgetMgr.getStringValue({
+        id: element.id,
+        formId: element.formId,
+      } as WidgetInfo)
+
+      if (initialWidgetValue) {
```

Comment:
nit: in order to reduce nesting and thus increase readability, it looks like you could do the reverse check:

```typescript
if (!initialWidgetValue) {
  return
}
```

Reply:
Updated 👍